snuba/clickhouse/formatter/query.py

```diff
@@ -37,11 +38,17 @@ def format_query(query: FormattableQuery) -> FormattedQuery:
     This is the entry point for any type of query, whether simple or
     composite.
     """
-    return FormattedQuery(_format_query_content(query, ClickhouseExpressionFormatter))
+    return FormattedQuery(
```

Comment:
Your current implementation doesn't handle complex expressions properly. Firstly, it doesn't sort all the parts of the query (group by, order by). Also because it is relying on string manipulation to sort expressions, it is going to improperly sort a lot of cases or just break the query. You are not handling nested conditions for example: 

```
"WHERE d = 1 AND c = 2 AND (b = 3 OR a = 4)" 
# becomes
["WHERE d = 1", "AND (b = 3 ", "AND c = 2 ", "OR a = 4)"]
```
This isn't just incorrectly sorted, it is also breaking the query by splitting the expression in the parentheses.

I don't think this is the best way of structuring this code. A cleaner way would be to extend `ClickhouseExpressionFormatter`, in a similar way that the  `ExpressionFormatterAnonymized` does. Then you could sort the expressions before passing them to the parent function to stringify. The `sort_fields` bool here would determine which formatter to use.

Reply:
Good point. I revised the sorting implementation to be conducted in the AST level. The query is now sorted based on an added attribute on the `ParsingContext` class. The actual sorting is now done at each recursive level when traversing the expressions in the FunctionCalls.

Additionally, since we are potentially sorting items of different types, the following ordering was created to ensure some level of consistency among classes: Literals < Columns < FunctionCalls (nested expressions). Note that we only sort string literals, else we default.